atlasdb-workload-server-api/src/main/java/com/palantir/atlasdb/workload/invariant/MismatchedValue.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * (c) Copyright 2023 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.workload.invariant;
+
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import java.util.Optional;
+import org.immutables.value.Value;
+
+@Value.Immutable
+public interface MismatchedValue {
+    @Value.Parameter
+    Optional<Integer> actual();
+
+    @Value.Parameter
+    Optional<Integer> expected();
+
+    @Value.Check
+    default void actualAndExpectedAreNotEqualToEachOther() {
+        Preconditions.checkArgument(
+                !actual().equals(expected()),
+                "Cannot construct MismatchedValues for two values which are equal.",
+                SafeArg.of("actual", actual()),
```

Comment:
kind of silly that we log both values, but I found it to be a better user experience, especially if we ever refactor this class to be generic (as it would be nice to see if it was due to reference equality or something else!)